### ISSUE TITLE
Garbage Collected StandardGpuResources Leads to Segmentation Fault in Python

### ISSUE THREAD

**Issue by josepablocam** [facebookresearch/faiss#505]
# Summary
When a `StandardGpuResources` is garbage collected, querying the index that was created using it produces a segmentation fault. This situation can easily arise if the index is created in a method and the resources object is a local variable, once it returns, that variable can be gargabe collected by Python, producing the issue described.

# Platform

OS: CentOS Linux release 7.5.1804 

Faiss version: faiss-gpu py36_cuda8.0.61_2

Faiss compilation options: 
Installed from conda

Running on:
- [ ] CPU
- [x] GPU

Interface: 
- [ ] C++
- [x] Python

# Reproduction instructions

Running the snippet below should reproduce the issue.

```
import faiss
import gc
import numpy as np

np.random.seed(42)
data = np.random.randn(10, 10).astype(np.float32)
res = faiss.StandardGpuResources()
index = faiss.GpuIndexFlatIP(res, data.shape[1])
index.add(data)
query = np.random.randn(2, 10).astype(np.float32)

# works nicely
index.search(query, 2)

# produces segfault once res is garbage collected
res = None
gc.collect()
index.search(query, 2)
```

**Comment by mdouze** [comment]
Hi 
See here for comments and workarounds: https://github.com/facebookresearch/faiss/wiki/Troubleshooting#crashes-in-pure-python-code

**Comment by josepablocam** [comment]
@mdouze Thanks. Totally missed this! Would it be reasonable to keep an issue open for this? (doesn't have to be this one, perhaps there's an older closed issue that would do). I searched through the issues first and didn't see anything, so assumed (incorrectly) it was unknown behavior. 

In either case, thanks for this useful link (and awesome tool).

**Comment by mdouze** [comment]
Fixed in 1.4.0
